### ISSUE TITLE
MLB XML API deprecation

### ISSUE THREAD

**Issue by ty-porter** [MLB-LED-Scoreboard/mlb-led-scoreboard#293]
It appears that MLB no longer supports its older GameDay XML API. The API wrapper used by the scoreboard [`mlbgame`](https://github.com/panzarino/mlbgame) depends on this API to render team standings.

There have been lots of reports of the scoreboard warning on startup `Failed to refresh standings.` due to this issue.

A sample API call to retrieve stats:
```
// URL
GET http://mlb.mlb.com/lookup/json/named.historical_standings_schedule_date.bam?season=2020&game_date=%272020/07/25%27&sit_code=%27h0%27&league_id=103&league_id=104&all_star_sw=%27N%27&version=48

// Response
'Check back soon!  Working on you request'
```

Short term, an appropriate fix may be to simply bypass the `mlbgame` dependency/old API only on the broken call and use the new [`MLB Data API`](https://appac.github.io/mlb-data-api-doc). 

Sample call to the new endpoint:

```
// URL
https://statsapi.mlb.com/api/v1/standings?season=2020&leagueId=103,104&date=07/25/2020&division=all

{
  "copyright": "Copyright 2020 MLB Advanced Media, L.P.  Use of any content on this page acknowledges agreement to the terms posted here http://gdx.mlb.com/components/copyright.txt",
  "records": [
    {
      "standingsType": "regularSeason",
      "league": {
        "id": 103,
        "link": "/api/v1/league/103"
      },
      "division": {
        "id": 200,
        "link": "/api/v1/divisions/200"
      },
      "sport": {
        "id": 1,
        "link": "/api/v1/sports/1"
      },
      "lastUpdated": "2020-07-25T18:27:02.000Z",
      "teamRecords": [
        {
          "team": {
            "id": 117,
            "name": "Houston Astros",
            "link": "/api/v1/teams/117"
          },
          "season": "2020",
          "streak": {
            "streakType": "wins",
            "streakNumber": 1,
            "streakCode": "W1"
          },
          "divisionRank": "1",
          "leagueRank": "3",
          "wildCardRank": "3",
          "sportRank": "6",
          "gamesPlayed": 1,
          "gamesBack": "-",
          "wildCardGamesBack": "-",
          "leagueGamesBack": "-",
          "springLeagueGamesBack": "-",
          "sportGamesBack": "0.5",
          "divisionGamesBack": "-",
          "conferenceGamesBack": "-",
          "leagueRecord": {
            "wins": 1,
            "losses": 0,
            "pct": "1.000"
          },
          . . . More teams . . .
     ]
  }
}
```

Long term it is most likely best to get rid of the `mlbgame` dependency and start using the appopriate API. It appears that the [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) package could be a suitable replacement candidate.

**Comment by pfightingpolish** [comment]
I was able to Frankenstein the [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) to make the standings work again. 

My code, as it has been throughout my journey, was an inelegant solution, so I'm not even going to try and share it in full. I've been heavily customizing my board and my Python is weak. I wasn't sure how to iterate through the teams and I've been lazy about coding back to the config.json file for the coordinates. I just hard-coded out the data five times, once for each position in the division, with mostly constants for the coordinates. Here's what first place looks like, if it helps anyone: 

```
      team_text = self.__team_shortener(self.data.current_standings().get('teams')[0].get('name'))
      graphics.DrawText(self.canvas, font["font"], 2, 22, self.team_name_color, team_text)

      record_text = "{}-{}".format(self.data.current_standings().get('teams')[0].get('w'), self.data.current_standings($
      if "-" in self.data.current_standings().get('teams')[0].get('gb'):
        gb_text = " -  "
      else:
        gb_text = format(self.data.current_standings().get('teams')[0].get('gb'))

      graphics.DrawText(self.canvas, font["font"], record_start, 22, self.team_stat_color, record_text)
      graphics.DrawText(self.canvas, font["font"], gb_start, 22, self.team_stat_color, gb_text)
```

`__team_shortener` is just a translation function that takes the full team name ("Houston Astros") and translates it to just the city "Houston"); I'm using a 64x128 board with a variable width font, so I can fit that on there as long as the Giants don't win or lose 100 games (which ain't happening this year either way). `get` is a powerful tool in the [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI). Changing `[0]` to 1, 2, 3 and 4 will give you the other places. 

Standings is a good gateway to [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI), as the standings work somewhat separately from the rest of the board. However, I'm going to warn that moving from [`mlbgame`](https://github.com/panzarino/mlbgame) to [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) appears to be a massive task that will require an almost complete redo of the code. I started poking around to try and see if I could get the current batter & pitcher names onto the scoreboard, which is very possible in the new API: 

```
print statsapi.get('game_linescore',{'gamePk':631642}).get('offense').get('batter')
print statsapi.get('game_linescore',{'gamePk':631642}).get('defense').get('pitcher').get('fullName')
```

But I realized quickly it wasn't going to be easy. The big thing is the difference in the key field used: [`mlbgame`](https://github.com/panzarino/mlbgame) works off the old game ID format (i.e. gid_2011_07_09_tbamlb_nyamlb_1) whereas the [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) works off the new one, where the ID is just an integer (631642 above). A healthy amount of the [`mlb-led-scoreboard`](https://github.com/MLB-LED-Scoreboard/mlb-led-scoreboard) code, at present, routinely splits up the old game ID and pulls info based off, say, the team abbreviation. The new one's going to require different translation in a bunch of places. I gave up on trying to translate the game ID to the gamePk being sought above. 

I'm not optimistic about any interoperability between the two, as it doesn't appear [`mlbgame`](https://github.com/panzarino/mlbgame) has been touched in about two years and has unanswered issues. I think we're more apt to just get `mlb-led-scoreboard` upgraded to 4.0. 

That said ... I was actually considering coming here and saying **THANK YOU, THANK YOU, THANK YOU** to all of you who have worked on this. I've had a fascination with scoreboards since the first time I went to a Brewers game as a five-year-old in 1987. This code has allowed me to recapture some of the awe and wonder. I've been obsessed from the moment I saw the Dodger example on Reddit. 

This code has had me trying to recreate the feeling of that old County Stadium board, familiar to anyone who saw the classic film "Major League" (made in Milwaukee). That's meant some custom fonts, really messing with the code and some late nights. But the results are ... pretty. 

![20200725_172952](https://user-images.githubusercontent.com/68492369/88467777-656db800-cea0-11ea-9dc6-6f8f276e3b4d.jpg)

![20200725_173007](https://user-images.githubusercontent.com/68492369/88467780-6a326c00-cea0-11ea-8763-6e291cd4efee.jpg)

![20200725_174256](https://user-images.githubusercontent.com/68492369/88467783-6c94c600-cea0-11ea-9a92-e8f30ce7833b.jpg)

![20200725_174323](https://user-images.githubusercontent.com/68492369/88467788-774f5b00-cea0-11ea-8e8d-a2ac5ff86a4f.jpg)

![20200725_175000](https://user-images.githubusercontent.com/68492369/88467791-7b7b7880-cea0-11ea-8825-a72f276b668f.jpg)

![image](https://user-images.githubusercontent.com/68492369/88467876-81258e00-cea1-11ea-9b2d-d043b1d23e30.png)

Having taken the code about as far as I think I can, the next task is the carpentry to build a housing. You can guess what I've got in mind.

**Comment by msroth** [comment]
Love your classic scoreboards.  Too bad the MLB-LED scoreboard code is broken due the discontinued XML API.  I just installed one of these in a really nice box in a friend's bar.  Imagine our disappointment.  At least the NHL scoreboard will work next month.

**Comment by pfightingpolish** [comment]
Games and headlines are still working. It's just the standings that are borked at the moment.

**Comment by pawptart** [comment]
> Love your classic scoreboards. Too bad the MLB-LED scoreboard code is broken due the discontinued XML API. I just installed one of these in a really nice box in a friend's bar. Imagine our disappointment. At least the NHL scoreboard will work next month.

Most of the XML API is still working, so anything that drives the actual game stats is still good for. It's just the standings API has apparently been migrated to StatsAPI. The fact that standings don't work doesn't break anything, but it is less interesting on off days.

**Comment by pawptart** [comment]
> However, I'm going to warn that moving from [`mlbgame`](https://github.com/panzarino/mlbgame) to [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) appears to be a massive task that will require an almost complete redo of the code.

Absolutely true. My understanding is the XML API is going away completely at some indeterminate point in the future.

**Comment by swemoney** [comment]
There was work being done last year to kind of get our own mlbgame library running using the json API. I haven't had a lot of time lately to take a look but I'll try and see what's working and what still needs work. The idea is to ultimately ditch the mlbgame library in favor of one that supports the json API. Just a question of if we have to write it ourselves or not.

**Comment by pawptart** [comment]
Just a cursory look at the codebase, seems like there are only a few API calls made through `mlbgame`:

```python
# The call this issue concerns
mlbgame.standings()

# Calls to grab today's game data
mlbgame.day()

# Not exactly sure what this returns, there may not be an analog for this on StatsAPI
mlbgame.overview()
```

There may be some child calls that I'm missing by a first pass but based on that I don't think writing our own library to support the JSON API is a bad idea.

**Comment by pfightingpolish** [comment]
Overview pulls in a lot of the info specific to a particular game, like team runs, hits, errors, the count and outs, etc. (though not current batter and pitcher, frustratingly). 

```
class Overview(mlbgame.object.Object):
    """Object to hold an overview of game information
    Properties:
        ampm
        attendance
        aw_lg_ampm
        away_ampm
        away_code
        away_division
        away_file_code
        away_games_back
        away_games_back_wildcard
        away_league_id
        away_loss
        away_name_abbrev
        away_preview_link
        away_probable_pitcher_era
        away_probable_pitcher_first
        away_probable_pitcher_first_name
        away_probable_pitcher_id
        away_probable_pitcher_last
        away_probable_pitcher_last_name
        away_probable_pitcher_losses
        away_probable_pitcher_name_display_roster
        away_probable_pitcher_number
        away_probable_pitcher_s_era
        away_probable_pitcher_s_losses
        away_probable_pitcher_s_wins
        away_probable_pitcher_stats_season
        away_probable_pitcher_stats_type
        away_probable_pitcher_throwinghand
        away_probable_pitcher_wins
        away_recap_link
        away_sport_code
        away_team_city
        away_team_errors
        away_team_hits
        away_team_id
        away_team_name
        away_team_runs
        away_time
        away_time_zone
        away_win
        balls
        date
        day
        double_header_sw
        elapsed_time
        first_pitch_et
        game_data_directory
        game_id
        game_nbr
        game_pk
        game_type
        gameday_link
        gameday_sw
        hm_lg_ampm
        home_ampm
        home_code
        home_division
        home_file_code
        home_games_back
        home_games_back_wildcard
        home_league_id
        home_loss
        home_name_abbrev
        home_preview_link
        home_probable_pitcher_era
        home_probable_pitcher_first
        home_probable_pitcher_first_name
        home_probable_pitcher_id
        home_probable_pitcher_last
        home_probable_pitcher_last_name
        home_probable_pitcher_losses
        home_probable_pitcher_name_display_roster
        home_probable_pitcher_number
        home_probable_pitcher_s_era
        home_probable_pitcher_s_losses
        home_probable_pitcher_s_wins
        home_probable_pitcher_stats_season
        home_probable_pitcher_stats_type
        home_probable_pitcher_throwinghand
        home_probable_pitcher_wins
        home_recap_link
        home_sport_code
        home_team_city
        home_team_errors
        home_team_hits
        home_team_id
        home_team_name
        home_team_runs
        home_time
        home_time_zone
        home_win
        id
        ind
        inning
        inning_state
        is_no_hitter
        is_perfect_game
        league
        location
        note
        official_scorer
        original_date
        outs
        photos_link
        preview
        scheduled_innings
        start_time
        status
        status_ind
        strikes
        tbd_flag
        tiebreaker_sw
        time
        time_aw_lg
        time_date
        time_date_aw_lg
        time_date_hm_lg
        time_hm_lg
        time_zone
        time_zone_aw_lg
        time_zone_hm_lg
        top_inning
        tv_station
        tz_aw_lg_gen
        tz_hm_lg_gen
        venue
        venue_id
        venue_name
        venue_w_chan_loc
        weather
        wind
        wrapup_link
    """
    pass
```

Day just returns a list of games on a given day. At present, it's how this project builds the list of games for the day. 

If we're working off [`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI), there's not a ton of pre-built functions, but there's the ability to dig deep into the dicts and get to endpoints using the `get()` command repeatedly. Different pieces might require hitting different endpoints, though. For instance, calling this: 

`statsapi.get('game_linescore',{'gamePk':631642})`

Produces this dict: 

`{u'innings': [{u'home': {u'leftOnBase': 2, u'runs': 1, u'errors': 0, u'hits': 1}, u'away': {u'leftOnBase': 2, u'runs': 0, u'errors': 0, u'hits': 2}, u'num': 1, u'ordinalNum': u'1st'}, {u'home': {u'leftOnBase': 1, u'runs': 0, u'errors': 0, u'hits': 0}, u'away': {u'leftOnBase': 2, u'runs': 1, u'errors': 0, u'hits': 2}, u'num': 2, u'ordinalNum': u'2nd'}, {u'home': {u'leftOnBase': 1, u'runs': 0, u'errors': 0, u'hits': 0}, u'away': {u'leftOnBase': 0, u'runs': 0, u'errors': 0, u'hits': 0}, u'num': 3, u'ordinalNum': u'3rd'}, {u'home': {u'leftOnBase': 2, u'runs': 0, u'errors': 0, u'hits': 2}, u'away': {u'leftOnBase': 1, u'runs': 2, u'errors': 0, u'hits': 2}, u'num': 4, u'ordinalNum': u'4th'}, {u'home': {u'leftOnBase': 0, u'runs': 2, u'errors': 0, u'hits': 2}, u'away': {u'leftOnBase': 1, u'runs': 1, u'errors': 0, u'hits': 1}, u'num': 5, u'ordinalNum': u'5th'}, {u'home': {u'leftOnBase': 0, u'runs': 0, u'errors': 0, u'hits': 0}, u'away': {u'leftOnBase': 0, u'runs': 2, u'errors': 0, u'hits': 1}, u'num': 6, u'ordinalNum': u'6th'}, {u'home': {u'leftOnBase': 3, u'runs': 0, u'errors': 0, u'hits': 1}, u'away': {u'leftOnBase': 0, u'runs': 2, u'errors': 0, u'hits': 3}, u'num': 7, u'ordinalNum': u'7th'}, {u'home': {u'leftOnBase': 0, u'runs': 0, u'errors': 0, u'hits': 0}, u'away': {u'leftOnBase': 1, u'runs': 0, u'errors': 0, u'hits': 0}, u'num': 8, u'ordinalNum': u'8th'}, {u'home': {u'leftOnBase': 0, u'runs': 0, u'errors': 0, u'hits': 0}, u'away': {u'leftOnBase': 0, u'runs': 0, u'errors': 0, u'hits': 0}, u'num': 9, u'ordinalNum': u'9th'}], u'balls': 0, u'currentInning': 9, u'inningHalf': u'Bottom', u'inningState': u'Bottom', u'currentInningOrdinal': u'9th', u'scheduledInnings': 9, u'strikes': 2, u'teams': {u'home': {u'leftOnBase': 9, u'runs': 3, u'errors': 0, u'hits': 6}, u'away': {u'leftOnBase': 7, u'runs': 8, u'errors': 0, u'hits': 11}}, u'defense': {u'onDeck': {u'fullName': u'Ben Gamel', u'link': u'/api/v1/people/592325', u'id': 592325}, u'catcher': {u'fullName': u'Omar Narvaez', u'link': u'/api/v1/people/553882', u'id': 553882}, u'center': {u'fullName': u'Lorenzo Cain', u'link': u'/api/v1/people/456715', u'id': 456715}, u'third': {u'fullName': u'Eric Sogard', u'link': u'/api/v1/people/519299', u'id': 519299}, u'shortstop': {u'fullName': u'Orlando Arcia', u'link': u'/api/v1/people/606115', u'id': 606115}, u'pitcher': {u'fullName': u'Eric Yardley', u'link': u'/api/v1/people/642233', u'id': 642233}, u'second': {u'fullName': u'Keston Hiura', u'link': u'/api/v1/people/669374', u'id': 669374}, u'right': {u'fullName': u'Avisail Garcia', u'link': u'/api/v1/people/541645', u'id': 541645}, u'batter': {u'fullName': u'Lorenzo Cain', u'link': u'/api/v1/people/456715', u'id': 456715}, u'inHole': {u'fullName': u'Orlando Arcia', u'link': u'/api/v1/people/606115', u'id': 606115}, u'team': {u'link': u'/api/v1/teams/158', u'id': 158, u'name': u'Milwaukee Brewers'}, u'left': {u'fullName': u'Ben Gamel', u'link': u'/api/v1/people/592325', u'id': 592325}, u'first': {u'fullName': u'Justin Smoak', u'link': u'/api/v1/people/475253', u'id': 475253}}, u'outs': 3, u'offense': {u'pitcher': {u'fullName': u'Dillon Maples', u'link': u'/api/v1/people/596027', u'id': 596027}, u'onDeck': {u'fullName': u'Kyle Schwarber', u'link': u'/api/v1/people/656941', u'id': 656941}, u'batter': {u'fullName': u'Javier Baez', u'link': u'/api/v1/people/595879', u'id': 595879}, u'inHole': {u'fullName': u'Willson Contreras', u'link': u'/api/v1/people/575929', u'id': 575929}, u'team': {u'link': u'/api/v1/teams/112', u'id': 112, u'name': u'Chicago Cubs'}}, u'isTopInning': False, u'copyright': u'Copyright 2020 MLB Advanced Media, L.P.  Use of any content on this page acknowledges agreement to the terms posted here http://gdx.mlb.com/components/copyright.txt'}`

This is a list of lists, so you can continue to call `get()` on it continually to dig deeper, as I did in this call: 

```
print statsapi.get('game_linescore',{'gamePk':631642}).get('offense').get('batter').get('fullName')
print statsapi.get('game_linescore',{'gamePk':631642}).get('defense').get('pitcher').get('fullName')
```

One function[`MLB-StatsAPI`](https://github.com/toddrob99/MLB-StatsAPI) does have is a fairly straightforward schedule parser. This call: 

`statsapi.schedule("2020-07-25")`

Produces this for every game (I edited this to only show two games, but you get the drift): 

`[{'status': u'Final', 'home_name': u'Chicago Cubs', 'game_datetime': u'2020-07-25T17:05:00Z', 'winning_pitcher': u'Brent Suter', 'save_pitcher': None, 'home_score': 3, 'winning_team': u'Chicago Cubs', 'game_id': 631642, 'summary': u'2020-07-25 - Milwaukee Brewers (8) @ Chicago Cubs (3) (Final)', 'away_probable_pitcher': u'Corbin Burnes', 'home_probable_pitcher': u'Yu Darvish', 'away_id': 158, 'venue_id': 17, 'venue_name': u'Wrigley Field', 'home_pitcher_note': u"Darvish dominated Milwaukee in '19, allowing one run with 15 strikeouts and one walk in two starts. In the second half, the righty had a 2.76 ERA with 118 strikeouts and seven walks in 81 2/3 innings.", 'game_type': u'R', 'game_num': 1, 'away_name': u'Milwaukee Brewers', 'home_id': 112, 'losing_team': u'Milwaukee Brewers', 'doubleheader': u'N', 'losing_pitcher': u'Yu Darvish', 'game_date': u'2020-07-25', 'away_score': 8, 'inning_state': u'Bottom', 'away_pitcher_note': u'Burnes had a rough stint in the starting rotation to open last season, when he allowed 11 home runs in four starts before the end of April. But he dazzled in Spring Training and Summer Camp.', 'current_inning': 9}, {'status': u'Final', 'home_name': u'Boston Red Sox', 'game_datetime': u'2020-07-25T17:35:00Z', 'winning_pitcher': u'Alex Cobb', 'save_pitcher': None, 'home_score': 2, 'winning_team': u'Boston Red Sox', 'game_id': 631666, 'summary': u'2020-07-25 - Baltimore Orioles (7) @ Boston Red Sox (2) (Final)', 'away_probable_pitcher': u'Alex Cobb', 'home_probable_pitcher': u'Martin Perez', 'away_id': 110, 'venue_id': 3, 'venue_name': u'Fenway Park', 'home_pitcher_note': u'The lefty was signed to be the No. 5 starter, but his responsibility has elevated greatly due to Chris Sale being out for the season and Eduardo Rodriguez opening on the injured list. P\xe9rez started fast (8-3 before the All-Star break) last year.', 'game_type': u'R', 'game_num': 1, 'away_name': u'Baltimore Orioles', 'home_id': 111, 'losing_team': u'Baltimore Orioles', 'doubleheader': u'N', 'losing_pitcher': u'Martin Perez', 'game_date': u'2020-07-25', 'away_score': 7, 'inning_state': u'Bottom', 'away_pitcher_note': u'Now 32, Cobb is in the third season of a four-year, $57 million contract. He made just three starts in 2019, missing most of the year to hip surgery. The veteran righty is 53-52 with a 3.85 ERA across parts of eight MLB seasons.', 'current_inning': 9},  ... ]`

Seeing this, I was able to create a list of games for a date: 

```
>>> i = 0
>>> while i < 15:
...  statsapi.schedule("2020-07-25")[i].get('game_id')
...  i += 1
...
631642
631666
631515
630992
630967
631108
630942
631241
631181
631453
631382
631565
631603
630899
631092
```

However, you need to somehow determine how many games are in a day, otherwise it throws an out of range error if you try to pull on a day when less than the whole league is playing. And, as you can see, it's not like game IDs have any rhyme or reason. 

It's possible to make this work, and even open up more possibilities for data to be shown on the board. It's late and I've literally only been working in Python for a few days, so I'm not going to solve all the riddles at the moment. But, maybe this'll help the folks working on making it happen.

**Comment by trevor-viljoen** [comment]
I think I actually wrote that section of code for mlbgame to pull the standings. I haven’t had time to dig into the new issue, but I did some work on writing python3 bindings for the json API a while back. Feel free to dig through the code, fork, make pull requests or suggestions. 

https://github.com/trevor-viljoen/mlbapi

**Comment by msroth** [comment]
> > Love your classic scoreboards. Too bad the MLB-LED scoreboard code is broken due the discontinued XML API. I just installed one of these in a really nice box in a friend's bar. Imagine our disappointment. At least the NHL scoreboard will work next month.
> 
> Most of the XML API is still working, so anything that drives the actual game stats is still good for. It's just the standings API has apparently been migrated to StatsAPI. The fact that standings don't work doesn't break anything, but it is less interesting on off days.

This hasn't been my experience.  I get nothing but day old game results.  No live game data at all.  I'm going to mess with my config, though nothing has changed since last season.

**Comment by pawptart** [comment]
> ```
> >>> i = 0
> >>> while i < 15:
> ...  statsapi.schedule("2020-07-25")[i].get('game_id')
> ...  i += 1
> ```

Just FYI, the Pythonic way to do this is a `for ... in` loop:
```
return [game.get('game_id') for game in statsapi.schedule("2020-07-25")]
```

Conveniently avoids having to hard code the number of teams.

**Comment by pfightingpolish** [comment]
> This hasn't been my experience. I get nothing but day old game results. No live game data at all. I'm going to mess with my config, though nothing has changed since last season.

Weird. Mine's chugging along just fine right now. 

Things I can think of that might cause old results: 

- Having Demo Date set to anything other than false in `config.json`
- Something's up with your day start time in `config.json`
- Your Pi could use a restart
- Your Pi has networking issues

Lots of other possibilities, too, but those are the first I can think of. 

Best of luck troubleshooting.

**Comment by pfightingpolish** [comment]
> Just FYI, the Pythonic way to do this is a `for ... in` loop:
> 
> ```
> return [game.get('game_id') for game in statsapi.schedule("2020-07-25")]
> ```
> 
> Conveniently avoids having to hard code the number of teams.

Yeah, I'm still getting my Pi sea legs. I will not be the expert coder here. Totally makes sense that way.

**Comment by msroth** [comment]
> > This hasn't been my experience. I get nothing but day old game results. No live game data at all. I'm going to mess with my config, though nothing has changed since last season.
> 
> Weird. Mine's chugging along just fine right now.
> 
> Things I can think of that might cause old results:
> 
> * Having Demo Date set to anything other than false in `config.json`
> * Something's up with your day start time in `config.json`
> * Your Pi could use a restart
> * Your Pi has networking issues
> 
> Lots of other possibilities, too, but those are the first I can think of.
> 
> Best of luck troubleshooting.

Thanks for the confirmation that games are still working and for the troubleshooting advice.  I had tried all those things except the "end of day" setting.  When I compared that section of the config file to the one in GitHub, I noticed there is a  time_format setting and the end of day is now "00:00".  Once those were changed it worked!  Thanks a million.

**Comment by pfightingpolish** [comment]
Just because there's no better place to post this, here's my finished product. 

It has over 1,000 likes in a 50,000+ Facebook group for Milwaukee nostalgia in just one day. I'm friends with a guy who worked with the real thing for many years, and he said both that he's impressed with it and that he's hoping to get a picture put in the control room for Miller Park's scoreboard (I am really tempted to build one for them, too). 

Thank you all so much for working on this. This thing feels like it's now my most prized possession. 

![image](https://user-images.githubusercontent.com/68492369/90472176-8adb9380-e0e5-11ea-9667-ec3b0d30d9df.png)

![Github Smaller](https://user-images.githubusercontent.com/68492369/90472292-c4ac9a00-e0e5-11ea-8a16-35b505ce425a.jpg)

**Comment by twfarley** [comment]
This is awesome! Are you planning to check in your changes to a fork?

**Comment by pfightingpolish** [comment]
I'd be willing to. I'd have to figure out how to do it. This is literally the first thing I've done involving Github. 

As I've said upthread, too, I will warn my code is, admittedly, somewhat disastrous. To get some of my changes, I translated [mlbgame](https://github.com/panzarino/mlbgame)'s game IDs into something that can pull the game IDs from [MLB-StatsAPI](https://github.com/toddrob99/MLB-StatsAPI), but to say my solutions were inelegant is putting it mildly. There were more than a few places where I probably should have referenced back to the `config.json` file, coordinates file or color file, but just hard-coded values in the interest of speed. My CS252 teacher from college would probably try and snap a ruler over my head for some of my code. 

Nonetheless, just like with my finagling the code, I can try and figure it out. I have some screen fatigue with all the social media hype right now, but I will remain active in this community and will try and create a fork when I get a chance.

**Comment by pfightingpolish** [comment]
> This is awesome! Are you planning to check in your changes to a fork?

I just created a fork to my profile and uploaded the files I changed. 

A quick outline of some of my differences. You would definitely not be able to run my code out of the box, but can certainly consult for some changes: 

- All the fonts are custom and variations on what was actually used on the County Stadium Omega board. Because the matrix directory on github is a direct link to hzeller's, I have my variant fonts in my main directory rather than in the matrix directory, where I have them housed on my install. They're just Fony modifications and the numbered sizes don't necessarily correspond to anything, as they're variable width.
 
- The whole board is in old-school amber. My install was 128x64 and gave a decent amount of flicker with the color, so I decided to go single-color for the whole thing. Take that into account. 

- Obviously the layout is different, but I think the big addition I made was to add current batter and pitcher to in-progress games. 

Yes, that pulls info from [MLBStats-API](https://github.com/toddrob99/MLB-StatsAPI), so there's also code to translate the game ID from [mlbgame](https://github.com/panzarino/mlbgame) to the gamePk in [MLBStats-API](https://github.com/toddrob99/MLB-StatsAPI). I've been poking around [MLBStats-API](https://github.com/toddrob99/MLB-StatsAPI) more, and I think I might have done this a much harder way than I had to, as it looks like there is a version of the old ID in the new 'game' endpoint. But there's also so much in `game` that I think it might be better and more efficient to narrow it down. 

I've started looking at trying to get play results to scroll, too, after they occur, but if I do that, I'd probably go through the `game-playByPlay` endpoint separately from the `game-linescore` endpoint where I get batter and pitcher rather than storing the general `game` endpoint, as then, you're storing every bit of StatCast data in a game — hot zones, pitchF/X, you name it — which might be too much for the Pi. 

- I also made the point to show a linescore with RHE (runs, hits and errors) rather than just each team and score. That still uses data from the old API, however. Probably only possible on w128 installs, but I think it's a useful addition. With that information on the board, I virtually turned off any no-hitter/perfect-game notifications by setting the number of innings before they display up to 20 (and if a team ever gets no-hit for 20 innings in a single game, I'll accept the scoreboard crashing). 

- I also fixed the standings by pulling from [MLBStats-API](https://github.com/toddrob99/MLB-StatsAPI). My ticker screen is personalized with different layout, team logo and MLB logo. I have mine set up for offdays to run the ticker, but cycle through the standings every five minutes. 

- I left a few of my "party trick" files in there, namely any of the files labeled `offdayXXXXX.py`. Those go to show that you can use your display to show other things, too. I will move around the offday files in the directory to do different stuff, like recreate the board from Rick Vaughan's entrance near the end of _Major League_ and posting pictures of friends with their name and stuff. Typically, I turn on the flag in `config.json` that makes it show the offday screen, then move the file that's doing what I want it to do into that spot rather than the ticker file, usually renaming the ticker file something like `offdayHOLD.py` while I'm doing my other schtuff. It gives you the chance to do some neat things that aren't out of line with some of the other, non-scoreboard things you can do with a matrix. 

Anyhow, check it out if you want to see some of the ways you can customize your install. 

I actually had a reporter from our local CBS affiliate in town over the other day to do a story on my board that should air in a couple weeks. I cited Github a bunch of times; we'll see if it makes the final cut.

**Comment by WardBrian** [comment]
Has there been any progress on this? I’m now getting errors on the non-existence of rawboxscore.xml from mlbgame, so I think it’s edged closer to the point of the XML api being completely defunct

**Comment by WardBrian** [comment]
I've re-written the data and standings to both use `statsapi` in my fork here: https://github.com/WardBrian/mlb-led-scoreboard/tree/stats-change

I've also made a bunch of other changes (including using py3, putting at-bat information in a similar way to pfightingpolish, and a few other items), so I'm not sure if it would work as a PR, but if anyone is curious it's available

**Comment by ty-porter** [comment]
Closed by #335
